doc/manual/methods.rst

```diff
@@ -440,6 +440,212 @@ can also constrain type parameters of methods::
+There is no general transform of one subtype into another subtype with a different parameter.
+(Quick review: do you see why this is?)
+
+The subtypes of ``AbstractArray`` typically implement two methods for expressing this:
+A method to convert the input to a subtype of a specific ``AbstractArray{T, N}`` abstract type;
+and a method to make a new uninitialized matrix with a specific element type.
+Sample implementations of these can be found in the standard library.
+Here is a basic example usage of them::
+
+    input = convert(AbstractArray{Eltype}, input)
+    output = similar(Eltype, input)
+
+As an extension of this, in cases where the algorithm wants to mutate the input array,
+``convert`` is insufficient as the return value may alias the original input.
+Combining ``similar`` (to make the output array) and ``copy!`` (to fill it with the input data)
+is a generic way to express the requirement for a mutable copy of the input argument::
+
+    copy_with_eltype(Eltype, input) = copy!(similar(Eltype, input), input)
+
+
+3. Iterated dispatch
+~~~~~~~~~~~~~~~~~~~~
+
+In order to dispatch a multi-level parametric argument list,
+often it is best to separate each level of dispatch into separate functions.
+This may sound similar in approach to single-dispatch, but as we shall see below, it is still more flexible.
+
+For example, trying to dispatch on the element-type of an array will often run into ambiguous situations.
+Instead, commonly code will dispatch first on the container type,
+then recurse down to a more specific method based on eltype.
+In most cases, the algorithms lend themselves conveniently to this hierarchical approach,
+while in other cases, this rigor must be resolved manually.
+This dispatching branching can be observed, for example, in the logic to sum two matrices::
+
+    # First dispatch selects the map algorithm for element-wise summation.
+    +(a::Matrix, b::Matrix) = map(+, a, b)
+    # Then dispatch handles each element and selects the appropriate
+    # common element type for the computation.
+    +(a, b) = +(promote(a, b)...)
+    # Once the elements have the same type, they can be added.
+    # For example, via primitive operations exposed by the processor.
+    +(a::Float, b::Float) = Core.add(a, b)
+
+
+4. Trait-based dispatch
+~~~~~~~~~~~~~~~~~~~~~~~
+
+A natural extension to the iterated dispatch above is to add a layer to
+method selection that identifies classes of types.
+We could do this by writing out a ``Union`` of the types with similar characteristics.
+But then this list would not be extensible.
+However, by expressing the property as a "trait",
+it is possible to express type behaviors in the abstract
+in a way that is flexible to new additions, but which has no performance impact.
+
+A trait is implemented by defining a pure generic function which returns a
+singleton value from a particular set, as a computation on the types of its arguments.
```

Comment:
This sentence could use an example/unpacking, perhaps tying into the `linearindexing`/`map` example below?